### PR TITLE
Only remove nulls from the middle of arrays

### PR DESCRIPTION
**Pull request by yufeih**
#3149

### CODE REVIEW

File: src/docfx/lib/JsonUtility.cs [hunk 1]
```diff
@@ -346,7 +357,7 @@ private static bool IsNullOrUndefined(this JToken token)
                 (token.Type == JTokenType.Undefined);
         }
 
-        private static void FindNulls(this JToken token, List<Error> errors, List<JToken> nullNodes)
+        private static void RemoveNullsCore(JToken token, List<Error> errors, List<JToken> nullNodes, List<JToken> removeNodes)
```
Comment (mingwli) [inline src/docfx/lib/JsonUtility.cs]: >private static void RemoveNullsCore(JToken token, List<Error> errors, Li [](start = 8, length = 72)

can this be a local function?

Comment (superyyrrzz) [inline src/docfx/lib/JsonUtility.cs]: >List<JToken> nullNodes, List<JToken> removeNodes [](start = 78, length = 48)

seems `nullNodes` is always identical to `removeNodes`, merge into one?

Comment (yufeih) [reply src/docfx/lib/JsonUtility.cs]: `nullNodes` contains nulls for object and array, `removeNodes` only contains nulls in arrays.

Comment (yufeih) [reply src/docfx/lib/JsonUtility.cs]: If could be, I was just trying to keep the existing convention here

File: docs/specs/metadata.yml [hunk 2]
```diff
@@ -9,6 +9,19 @@ outputs:
   docs/a.json: |
     { "metadata": { "robots": "noindex, nofollow" } }
 ---
+# Preserve null custom metadata
```
[the remark this answers is not in the document]
Comment (yufeih) [reply docs/specs/metadata.yml]: @live1206 , is there an existing test that covers this?

[the remark this answers is not in the document]
Comment (live1206) [reply docs/specs/metadata.yml]: https://github.com/dotnet/docfx/blob/bc85753222af5147b53e265b1ff77908bdf3f87e/test/docfx.Test/lib/JsonUtilityTest.cs#L225

**Pull request merged**